File: 20251220 - UI_MAP_BUILDER_INTEGRATION/solution_first/placement_calculator.py
"""
Placement Calculator for Solution-First Placement

Converts matched patterns into actual item coordinates.
"""

from typing import List, Dict, Tuple, Optional
from dataclasses import dataclass
from .pattern_library import Pattern


Coord = Tuple[int, int, int]


@dataclass
class ItemPlacement:
    """A single item placement with metadata."""
    coord: Coord
    item_type: str
    pattern_id: str
    segment_idx: int
    repeat: int


class PlacementCalculator:
    """
    Calculates item positions from patterns and segment coordinates.
    """
    
    def calculate_for_segment(
        self, 
        segment_coords: List[Coord], 
        pattern: Pattern, 
        segment_idx: int
    ) -> List[ItemPlacement]:
        """
        Calculate item placements for a single segment.
        
        Args:
            segment_coords: List of coordinates in the segment
            pattern: The pattern to apply
            segment_idx: Index of this segment (for tracking)
        
        Returns:
            List of ItemPlacement objects
        """
        placements = []
        pattern_length = pattern.length
        segment_length = len(segment_coords)
        
        # How many complete pattern repeats fit?
        num_repeats = segment_length // pattern_length
        
        for repeat in range(num_repeats):
            base_idx = repeat * pattern_length
            
            for offset, item_type in zip(pattern.item_coord_offsets, pattern.item_types):
                coord_idx = base_idx + offset
                
                if coord_idx < segment_length:
                    placements.append(ItemPlacement(
                        coord=segment_coords[coord_idx],
                        item_type=item_type,
                        pattern_id=pattern.id,
                        segment_idx=segment_idx,
                        repeat=repeat
                    ))
        
        return placements
    
    def calculate_for_all_segments(
        self,
        segments: List[List[Coord]],
        pattern_matches: List[Tuple[int, Pattern]]
    ) -> List[ItemPlacement]:
        """
        Calculate placements for all segments.
        
        Args:
            segments: List of segment coordinate lists
            pattern_matches: List of (segment_idx, pattern) tuples
        
        Returns:
            All item placements across all segments
        """
        all_placements = []
        
        for seg_idx, pattern in pattern_matches:
            if seg_idx < len(segments):
                segment_coords = segments[seg_idx]
                placements = self.calculate_for_segment(
                    segment_coords, pattern, seg_idx
                )
                all_placements.extend(placements)
        
        return all_placements
    
    def filter_invalid_placements(
        self,
        placements: List[ItemPlacement],
        start_pos: Coord,
        target_pos: Coord,
        path_coords: List[Coord] = None
    ) -> List[ItemPlacement]:
        """
        Filter out placements at invalid positions (start, goal, off-path).
        
        Args:
            placements: List of ItemPlacements
            start_pos: Start position to exclude
            target_pos: Goal position to exclude
            path_coords: Optional path coords for on-path check
        
        Returns:
            Filtered list of valid placements
        """
        path_set = set(path_coords) if path_coords else None
        
        valid = []
        for p in placements:
            # Skip start/goal
            if p.coord == start_pos or p.coord == target_pos:
                continue
            
            # Skip off-path if path provided
            if path_set and p.coord not in path_set:
                continue
            
            valid.append(p)
        
        return valid

    def to_item_dicts(self, placements: List[ItemPlacement]) -> List[Dict]:
        """
        Convert ItemPlacement objects to simple dicts for map assembly.
        """
        return [
            {
                'type': p.item_type,
                'pos': p.coord,
                'position': list(p.coord),  # Some code expects list format
                'pattern_id': p.pattern_id,
                'segment_idx': p.segment_idx
            }
            for p in placements
        ]
    
    def verify_placements(
        self,
        placements: List[ItemPlacement],
        path_coords: List[Coord],
        start_pos: Coord,
        target_pos: Coord
    ) -> Tuple[bool, List[str]]:
        """
        Verify that all placements are valid.
        
        Checks:
        1. All items on path
        2. No duplicate coords
        3. No items at start/goal
        4. Diversity check (≥2 item types)
        
        Returns:
            (success, list of error messages)
        """
        errors = []
        path_set = set(path_coords)
        
        # Check 1: All items on path
        for p in placements:
            if p.coord not in path_set:
                errors.append(f"Item at {p.coord} not on path")
        
        # Check 2: No duplicates
        coords = [p.coord for p in placements]
        if len(coords) != len(set(coords)):
            errors.append("Duplicate item coordinates found")
        
        # Check 3: No items at start/goal
        for p in placements:
            if p.coord == start_pos:
                errors.append("Item at start position")
            if p.coord == target_pos:
                errors.append("Item at goal position")
        
        # Check 4: Diversity
        item_types = set(p.item_type for p in placements)
        if len(placements) > 0 and len(item_types) < 2:
            errors.append(f"Low item diversity: only {item_types}")
        
        return (len(errors) == 0, errors)

File: apps/map-builder-app/src/map-generator/handlers/PlacementCalculator.ts
/**
 * Placement Calculator
 * 
 * Calculates item positions from patterns and segments.
 * Ported from Python: placement_calculator.py
 */

import { Coord } from '../types';
import { Pattern } from './PatternLibrary';
import { ItemPlacement as PedagogicalItemPlacement } from './PedagogicalStrategyHandler';

export interface ItemPlacement {
    type: string;
    pos: Coord;
    pattern_id: string;
    segment_idx: number;
    offset_in_segment: number;
}

interface PatternMatch {
    segmentIdx: number;
    pattern: Pattern;
}

/**
 * Placement Calculator
 * Handles position calculations for pattern-based placement.
 */
export class PlacementCalculator {
    
    /**
     * Calculate placements for all segments based on pattern matches.
     */
    calculateForAllSegments(
        segments: Coord[][],
        patternMatches: PatternMatch[]
    ): ItemPlacement[] {
        const placements: ItemPlacement[] = [];

        for (const match of patternMatches) {
            const segment = segments[match.segmentIdx];
            if (!segment || segment.length === 0) continue;

            const segmentPlacements = this.calculateForSegment(
                segment,
                match.pattern,
                match.segmentIdx
            );
            placements.push(...segmentPlacements);
        }

        return placements;
    }

    /**
     * Calculate placements for a single segment.
     */
    calculateForSegment(
        segment: Coord[],
        pattern: Pattern,
        segmentIdx: number
    ): ItemPlacement[] {
        const placements: ItemPlacement[] = [];
        const segmentLength = segment.length;

        // Scale offsets to fit segment
        const scaleFactor = segmentLength / pattern.min_segment_length;

        for (let i = 0; i < pattern.item_coord_offsets.length; i++) {
            const offset = pattern.item_coord_offsets[i];
            const itemType = pattern.item_types[i % pattern.item_types.length];

            // Calculate actual position in segment
            let scaledOffset = Math.floor(offset * scaleFactor);
            scaledOffset = Math.min(scaledOffset, segmentLength - 1);
            scaledOffset = Math.max(scaledOffset, 0);

            if (scaledOffset < segment.length) {
                placements.push({
                    type: itemType,
                    pos: segment[scaledOffset],
                    pattern_id: pattern.id,
                    segment_idx: segmentIdx,
                    offset_in_segment: scaledOffset
                });
            }
        }

        return placements;
    }

    /**
     * Filter out invalid placements (start, target, off-path).
     */
    filterInvalidPlacements(
        placements: ItemPlacement[],
        startPos: Coord,
        targetPos: Coord,
        pathCoords: Coord[]
    ): ItemPlacement[] {
        const pathSet = new Set(pathCoords.map(c => this.coordToKey(c)));
        const startKey = this.coordToKey(startPos);
        const targetKey = this.coordToKey(targetPos);

        return placements.filter(p => {
            const key = this.coordToKey(p.pos);
            // Not at start or target
            if (key === startKey || key === targetKey) return false;
            // Must be on path
            if (!pathSet.has(key)) return false;
            return true;
        });
    }

    /**
     * Verify placements meet requirements.
     */
    verifyPlacements(
        placements: ItemPlacement[],
        pathCoords: Coord[],
        startPos: Coord,
        targetPos: Coord
    ): { success: boolean; errors: string[] } {
        const errors: string[] = [];

        if (placements.length === 0) {
            errors.push('No placements generated');
        }

        // Check for duplicates
        const posSet = new Set<string>();
        for (const p of placements) {
            const key = this.coordToKey(p.pos);
            if (posSet.has(key)) {
                errors.push(`Duplicate placement at ${key}`);
            }
            posSet.add(key);
        }

        // Minimum items check
        if (placements.length < 3) {
            errors.push(`Too few placements: ${placements.length} (min: 3)`);
        }

        return {
            success: errors.length === 0,
            errors
        };
    }

    /**
     * Convert placements to item dicts for layout.
     */
    toItemDicts(placements: ItemPlacement[]): PedagogicalItemPlacement[] {
        return placements.map(p => ({
            type: p.type,
            pos: p.pos,
            position: [...p.pos],
            pattern_id: p.pattern_id,
            segment_idx: p.segment_idx
        }));
    }

    /**
     * Remove duplicate positions, keeping first occurrence.
     */
    removeDuplicates(placements: ItemPlacement[]): ItemPlacement[] {
        const seen = new Set<string>();
        return placements.filter(p => {
            const key = this.coordToKey(p.pos);
            if (seen.has(key)) return false;
            seen.add(key);
            return true;
        });
    }

    private coordToKey(coord: Coord): string {
        return `${coord[0]},${coord[1]},${coord[2]}`;
    }
}

// Singleton instance
let calculatorInstance: PlacementCalculator | null = null;

export function getPlacementCalculator(): PlacementCalculator {
    if (!calculatorInstance) {
        calculatorInstance = new PlacementCalculator();
    }
    return calculatorInstance;
}
